src/main/java/uk/ac/ebi/eva/contigalias/datasource/ENAAssemblyDataSource.java

```diff
@@ -0,0 +1,108 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Repository("ENADataSource")
+public class ENAAssemblyDataSource implements AssemblyDataSource {
+
+    private final ENABrowserFactory factory;
+
+    private final ENAAssemblyReportReaderFactory readerFactory;
+
+    @Autowired
+    public ENAAssemblyDataSource(ENABrowserFactory factory,
+                                 ENAAssemblyReportReaderFactory readerFactory) {
+        this.factory = factory;
+        this.readerFactory = readerFactory;
+    }
+
+    @Override
+    public Optional<AssemblyEntity> getAssemblyByAccession(String accession) throws IOException {
+        ENABrowser enaBrowser = factory.build();
+        enaBrowser.connect();
+
+        AssemblyEntity assemblyEntity;
+        try (InputStream stream = enaBrowser.getAssemblyReportInputStream(accession)) {
+            ENAAssemblyReportReader reader = readerFactory.build(stream);
+            assemblyEntity = reader.getAssemblyEntity();
+        } finally {
+            enaBrowser.disconnect();
+        }
+        return Optional.of(assemblyEntity);
+    }
+
+    /**
+     * Adds ENA sequence names to chromosomes and scaffolds in an assembly.
+     *
+     * @param optional {@link AssemblyEntity} to get ENA sequence names for
+     * @throws IOException Passes IOException thrown by {@link #getAssemblyByAccession(String)}
+     */
+    public void getENASequenceNamesForAssembly(Optional<AssemblyEntity> optional) throws IOException {
+        if (optional.isPresent()) {
+            AssemblyEntity targetAssembly = optional.get();
+            if (!hasAllEnaSequenceNames(targetAssembly)) {
+                String genbank = targetAssembly.getGenbank();
+                Optional<AssemblyEntity> enaAssembly = getAssemblyByAccession(genbank);
+
+                if (enaAssembly.isPresent()) {
+                    AssemblyEntity sourceAssembly = enaAssembly.get();
+                    putENASequenceNames(sourceAssembly.getChromosomes(), targetAssembly.getChromosomes());
+                    putENASequenceNames(sourceAssembly.getScaffolds(), targetAssembly.getScaffolds());
+                }
+            }
+        }
+    }
+
+    public boolean hasAllEnaSequenceNames(AssemblyEntity assembly) {
+        List<ChromosomeEntity> chromosomes = assembly.getChromosomes();
+        List<ScaffoldEntity> scaffolds = assembly.getScaffolds();
+        return Stream.concat(chromosomes.stream(), scaffolds.stream())
+                     .allMatch(sequence -> sequence.getEnaSequenceName() != null);
+    }
+
+    private void putENASequenceNames(
+            List<? extends SequenceEntity> sourceSequences, List<? extends SequenceEntity> targetSequences) {
+        Stream.concat(targetSequences.stream(), sourceSequences.stream())
+              .collect(Collectors.toMap(SequenceEntity::getGenbank, Function.identity(),
```

Comment:
**Judgment call:** This is a pretty clever use of lambdas but suggest re-writing in an imperative style since there are a lot of things going on.

Reply:
Re-wrote with good old `for`-loops...